core/templates/ElemUtils.scala

```diff
@@ -0,0 +1,193 @@
+                                        len: Int,
+                                        s: CharSequence) = {
+    var i = i0
+    while (
+      if (i >= len) false
+      else {
+        val c2 = s.charAt(i)
+        if (c2 < ' ' | c2 == '"' | c2 == '\\') false
+        else {
+          arr(arrOffset + i - i0) = c2.toElem
+          i += 1
+          true
+        }
+      }
+    ) ()
+
+    i - 1
+  }
+
+  private def appendSimpleStringSectionNoUnicode0(arr: Array[Elem],
+                                                 arrOffset: Int,
+                                                 i0: Int,
+                                                 len: Int,
+                                                 s: CharSequence) = {
+    var i = i0
+    while (
+      if (i >= len) false
+      else {
+        val c2 = s.charAt(i)
+        if (c2 < ' ' || c2 > 127 || c2 == '"' || c2 == '\\') false
+        else {
+          arr(arrOffset + i - i0) = c2.toElem
+          i += 1
+          true
+        }
+      }
+    ) ()
+
+    i - 1
+  }
+
+
+  def parseIntegralNum(arr: Array[Elem], arrOffset: Int, arrLength: Int, decIndex: Int, expIndex: Int) = {
+    val expMul =
+      if (expIndex == -1) 1
+      else {
+        var mult = 1
+        val e = parseLong(arr, arrOffset + expIndex + 1, arrOffset + arrLength)
+        var i = 0
+        while (i < e) {
+          if (mult >= Long.MaxValue / 10) throw new Abort("expected integer")
+          mult = mult * 10
+          i += 1
+        }
+        mult
+      }
+
+    val intPortion = {
+      val end =
+        if (decIndex != -1) decIndex
+        else if (expIndex != -1) expIndex
+        else arrLength
+
+      parseLong(arr, arrOffset, arrOffset + end) * expMul
+    }
+
+    val decPortion =
+      if (decIndex == -1) 0
+      else {
+        val end = if (expIndex != -1) expIndex else arrLength
+        var value = parseLong(arr, arrOffset + decIndex + 1, arrOffset + end) * expMul
+        var i = end - (decIndex + 1)
+        while (i > 0) {
+          value = value / 10
+          i -= 1
+        }
+        if (arr(arrOffset) == '-') -value else value
+      }
+
+    intPortion + decPortion
+  }
+
+  def parseLong(cs0: Array[Elem], start0: Int, end0: Int): Long = {
+    if ((start0 | end0 | end0 - start0 | (cs0.length - end0)) < 0) throw new IndexOutOfBoundsException
+
+    // If we do not copy the data from `cs0` into our own local array before
+    // parsing it, we take a significant performance penalty in the
+    // `integers Read` benchmarks, but *only* when run together with the rest
+    // of the benchmarks! When `integers Read` isrun alone, this does not
+    // happen, and is presumably something to do with the JIT compiler.
+
+    // Since any real world use case would exercise all sorts of code paths,
+    // it would more closely resemble the "all benchmarks together" case
+    // rather, and so we leave this copy in-place to optimize performance
+    // for that scenario.
```

Comment:
This is especially weird behavior that I don't understand. @plokhotnyuk have you seen anything like this before in your own performance work? I'm aware of things like the JIT causing method calls to slow down when running a bunch of code paths cause callsites to become megamorphic, but I've never heard of JIT-related slowdowns when reading and writing entries directly into arrays like this...

Reply:
Both way of testing are suitable to understand the parser work: the separated one to see how parsing of some data type works in isolation, and the combined one to see more realistic scenarios that usually based on real-world samples.

From the start of development jsoniter-scala doesn't have any predefined codecs and as much as possible use `private[this]` methods for generated codecs, so that the type-class interface is generated only for the top-level class of the nested data structures.

Reply:
@plokhotnyuk the unusual thing is that in both cases, the benchmark runs the parser in a tight single-threaded loop again and again in isolation (nothing else running in the JVM/OS).

The difference is that when the benchmark is run on a "clean" JVM right after startup, this code is fast. When the JVM is run on a "dirty" JVM that has already run a bunch of other benchmarks on the same thread (but which have all since terminated), this code is slow. That's the part that's rather unusual